In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
df=pd.read_csv('/content/Global_Pollution_Analysis.csv')
df.head()

Country  Year  Air_Pollution_Index  Water_Pollution_Index  \
0       Hungary  2005               272.70                 124.27   
1     Singapore  2001                86.72                  60.34   
2       Romania  2016                91.59                  83.36   
3  Cook Islands  2018               280.61                  67.16   
4      Djibouti  2008               179.16                 127.53   

   Soil_Pollution_Index  Industrial_Waste (in tons)  \
0                 51.95                    94802.83   
1                117.22                    56283.92   
2                121.72                    56256.02   
3                 93.58                    74864.73   
4                121.55                    76862.06   

   Energy_Recovered (in GWh)  CO2_Emissions (in MT)  Renewable_Energy (%)  \
0                     158.14                   5.30                 41.11   
1                     498.04                   6.34                 36.44   
2                     489.51                  49.69                  9.38   
3                     145.18                   8.91                 18.97   
4                      40.38                  14.93                 34.00   

   Plastic_Waste_Produced (in tons)  Energy_Consumption_Per_Capita (in MWh)  \
0                          37078.88                                   12.56   
1                          33128.20                                    5.23   
2                          18803.46                                   13.15   
3                           9182.27                                    0.78   
4                          39235.12                                   12.84   

   Population (in millions)  GDP_Per_Capita (in USD)  
0                     42.22                 20972.96  
1                    137.25                 34850.41  
2                    124.47                 57773.15  
3                     67.80                 21837.51  
4                    186.52                 41379.37

In [ ]:
df=df.dropna(how='all')
df.head()

Country  Year  Air_Pollution_Index  Water_Pollution_Index  \
0       Hungary  2005               272.70                 124.27   
1     Singapore  2001                86.72                  60.34   
2       Romania  2016                91.59                  83.36   
3  Cook Islands  2018               280.61                  67.16   
4      Djibouti  2008               179.16                 127.53   

   Soil_Pollution_Index  Industrial_Waste (in tons)  \
0                 51.95                    94802.83   
1                117.22                    56283.92   
2                121.72                    56256.02   
3                 93.58                    74864.73   
4                121.55                    76862.06   

   Energy_Recovered (in GWh)  CO2_Emissions (in MT)  Renewable_Energy (%)  \
0                     158.14                   5.30                 41.11   
1                     498.04                   6.34                 36.44   
2                     489.51                  49.69                  9.38   
3                     145.18                   8.91                 18.97   
4                      40.38                  14.93                 34.00   

   Plastic_Waste_Produced (in tons)  Energy_Consumption_Per_Capita (in MWh)  \
0                          37078.88                                   12.56   
1                          33128.20                                    5.23   
2                          18803.46                                   13.15   
3                           9182.27                                    0.78   
4                          39235.12                                   12.84   

   Population (in millions)  GDP_Per_Capita (in USD)  
0                     42.22                 20972.96  
1                    137.25                 34850.41  
2                    124.47                 57773.15  
3                     67.80                 21837.51  
4                    186.52                 41379.37

In [ ]:
from sklearn.preprocessing import MinMaxScaler
scaler=MinMaxScaler()
df["Industrial_Waste (in tons)"]=scaler.fit_transform(df[["Industrial_Waste (in tons)"]])
df["CO2_Emissions (in MT)"]=scaler.fit_transform(df[["CO2_Emissions (in MT)"]])
pollution_indices = ['Air_Pollution_Index', 'Water_Pollution_Index', 'Soil_Pollution_Index']
df[pollution_indices] = scaler.fit_transform(df[pollution_indices])
df.head()

Country  Year  Air_Pollution_Index  Water_Pollution_Index  \
0       Hungary  2005             0.898042               0.553778   
1     Singapore  2001             0.147062               0.173673   
2       Romania  2016             0.166727               0.310542   
3  Cook Islands  2018             0.929982               0.214222   
4      Djibouti  2008             0.520331               0.573161   

   Soil_Pollution_Index  Industrial_Waste (in tons)  \
0              0.295481                    0.949995   
1              0.768178                    0.559811   
2              0.800768                    0.559529   
3              0.596973                    0.748028   
4              0.799537                    0.768261   

   Energy_Recovered (in GWh)  CO2_Emissions (in MT)  Renewable_Energy (%)  \
0                     158.14               0.070756                 41.11   
1                     498.04               0.092527                 36.44   
2                     489.51               1.000000                  9.38   
3                     145.18               0.146326                 18.97   
4                      40.38               0.272347                 34.00   

   Plastic_Waste_Produced (in tons)  Energy_Consumption_Per_Capita (in MWh)  \
0                          37078.88                                   12.56   
1                          33128.20                                    5.23   
2                          18803.46                                   13.15   
3                           9182.27                                    0.78   
4                          39235.12                                   12.84   

   Population (in millions)  GDP_Per_Capita (in USD)  
0                     42.22                 20972.96  
1                    137.25                 34850.41  
2                    124.47                 57773.15  
3                     67.80                 21837.51  
4                    186.52                 41379.37

In [ ]:
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()
df["Country"]=le.fit_transform(df["Country"])
df["Year"]=le.fit_transform(df["Year"])
df.head()

Country  Year  Air_Pollution_Index  Water_Pollution_Index  \
0       77     5             0.898042               0.553778   
1      147     1             0.147062               0.173673   
2      136    16             0.166727               0.310542   
3       38    18             0.929982               0.214222   
4       46     8             0.520331               0.573161   

   Soil_Pollution_Index  Industrial_Waste (in tons)  \
0              0.295481                    0.949995   
1              0.768178                    0.559811   
2              0.800768                    0.559529   
3              0.596973                    0.748028   
4              0.799537                    0.768261   

   Energy_Recovered (in GWh)  CO2_Emissions (in MT)  Renewable_Energy (%)  \
0                     158.14               0.070756                 41.11   
1                     498.04               0.092527                 36.44   
2                     489.51               1.000000                  9.38   
3                     145.18               0.146326                 18.97   
4                      40.38               0.272347                 34.00   

   Plastic_Waste_Produced (in tons)  Energy_Consumption_Per_Capita (in MWh)  \
0                          37078.88                                   12.56   
1                          33128.20                                    5.23   
2                          18803.46                                   13.15   
3                           9182.27                                    0.78   
4                          39235.12                                   12.84   

   Population (in millions)  GDP_Per_Capita (in USD)  
0                     42.22                 20972.96  
1                    137.25                 34850.41  
2                    124.47                 57773.15  
3                     67.80                 21837.51  
4                    186.52                 41379.37

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, classification_report
columns=["Air_Pollution_Index","Water_Pollution_Index",'Soil_Pollution_Index','Industrial_Waste (in tons)']
y=df["CO2_Emissions (in MT)"]
bins = [y.min() - 1, y.quantile(0.33), y.quantile(0.66), y.max() + 1]
labels = ['Low', 'Medium', 'High']
df['Pollution_Severity'] = pd.cut(df['CO2_Emissions (in MT)'], bins=bins, labels=labels)
x=df[columns]
y=df["Pollution_Severity"]
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)
gnb = MultinomialNB()
gnb.fit(x_train, y_train)
y_pred = gnb.predict(x_test)
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy (exact matches): {accuracy}")
classification_report = classification_report(y_test, y_pred)
print("Classification Report:")
print(classification_report)
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)
print("\nConfusion Matrix:")
print(cm)

Accuracy (exact matches): 0.3
Classification Report:
              precision    recall  f1-score   support

        High       0.30      1.00      0.46        12
         Low       0.00      0.00      0.00        14
      Medium       0.00      0.00      0.00        14

    accuracy                           0.30        40
   macro avg       0.10      0.33      0.15        40
weighted avg       0.09      0.30      0.14        40


Confusion Matrix:
[[12  0  0]
 [14  0  0]
 [14  0  0]]


/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


In [ ]:
from sklearn.neighbors import KNeighborsClassifier
knn=KNeighborsClassifier(n_neighbors=4)
knn.fit(x_train,y_train)
y_pred=knn.predict(x_test)
accuracy=accuracy_score(y_test,y_pred)
print(f"Accuracy (exact matches): {accuracy}")
# classification_report_str=classification_report(y_test,y_pred)
# print("Classification Report:")
# print(classification_report_str)
from sklearn.metrics import confusion_matrix
cm=confusion_matrix(y_test,y_pred)
print("\nConfusion Matrix:")
print(cm)

Accuracy (exact matches): 0.4

Confusion Matrix:
[[7 3 2]
 [4 8 2]
 [7 6 1]]


In [ ]:
from sklearn.tree import DecisionTreeClassifier
dtc=DecisionTreeClassifier(max_depth=4,min_samples_split=50)
dtc.fit(x_train,y_train)
y_pred=dtc.predict(x_test)
accuracy=accuracy_score(y_test,y_pred)
print(f"Accuracy (exact matches): {accuracy}")
cm=confusion_matrix(y_test,y_pred)
print("\nConfusion Matrix:")
print(cm)

Accuracy (exact matches): 0.325

Confusion Matrix:
[[ 1 10  1]
 [ 0 11  3]
 [ 1 12  1]]


## Summary of Global Pollution Analysis

This analysis explored a dataset on global pollution, focusing on air, water, and soil pollution indices, industrial waste, and CO2 emissions.

**Data Preprocessing:**
The data was loaded and missing values were handled by dropping rows with all missing values. Several features, including pollution indices, industrial waste, and CO2 emissions, were scaled using MinMaxScaler to bring them to a similar range. The 'Country' and 'Year' columns were label encoded for use in the models.

**Pollution Severity Classification:**
To classify pollution severity, the 'CO2_Emissions (in MT)' was discretized into three categories: 'Low', 'Medium', and 'High' based on quantile values.

**Model Evaluation:**
Three classification models were implemented and evaluated:

*   **Multinomial Naive Bayes:** Achieved an accuracy of 0.30. The model primarily predicted the 'High' class, indicating potential issues with data imbalance or feature suitability for this model.
*   **K-Nearest Neighbors:** Achieved an accuracy of 0.40 (with n_neighbors=4). Similar to Naive Bayes, the model's performance was low, suggesting challenges in distinguishing between pollution severity levels.I got the best results with 4.
*   **Decision Tree Classifier:** Achieved an accuracy of 0.325 (with max_depth=4, min_samples_split=50). While slightly better than the other models, the accuracy is still low. The decision tree visualization provided some insight into how the model makes splits based on features.

**To improve the classification model, it is recommended to:**

1.  **Investigate Class Distribution:** Analyze the distribution of 'Low', 'Medium', and 'High' pollution severity classes to identify and address potential data imbalance using techniques like resampling.
2.  **Explore Feature Engineering and Selection:** Identify and incorporate other potentially relevant features from the dataset, or engineer new features that could improve the predictive power of the models.
3.  **Perform Hyperparameter Tuning:** Optimize the hyperparameters of the chosen models to potentially improve their performance.


The best performance was given by KNN, but it is still very low accuracy, so we can conclude that these models are not very good for this classification and we have to apply above given measures to improve the accuracy.